base/src/main/java/org/mozilla/jss/pkcs12/MacData.java

```diff
@@ -115,18 +141,31 @@ public MacData( Password password, byte[] macSalt,
 
         try {
             // generate key from password and salt
-            KeyGenerator kg = token.getKeyGenerator(KeyGenAlgorithm.PBA_SHA1_HMAC);
+            if(algID == null) {
+                algID = new AlgorithmIdentifier(DigestAlgorithm.SHA1.toOID());
+            }
+            KeyGenerator kg = null;
+            JSSMessageDigest digest = null;
+            if(DigestAlgorithm.SHA1.toOID().equals(algID.getOID())){
+                kg = token.getKeyGenerator(KeyGenAlgorithm.PBA_SHA1_HMAC);
+                digest = token.getDigestContext(HMACAlgorithm.SHA1);
+            }
+            if(DigestAlgorithm.SHA256.toOID().equals(algID.getOID())){
```

Comment:
Are we going to want to add one for 384 as well?

Reply:
I am not sure about the adoption but the plan should be to add all the option to 512 in future update.

Reply:
Makes sense.